### PR TITLE
Sentry fix: Check for app version only if user is logged in

### PR DESCRIPTION
**Pull request by yksolanki9**
- Fixes https://sentry.io/organizations/fyle-technologies-private-limi/issues/2234550581/?project=5622998 and https://sentry.io/organizations/fyle-technologies-private-limi/issues/2244675010/?project=5622998
- Moved `checkAppSupportedVersion()` call to `platform.ready()` as device plugin may not always be initialized inside `ngOnInit()`
- Some refactoring

### CODE REVIEW

File: src/app/app.component.ts [hunk 1]
```diff
@@ -136,43 +137,22 @@ export class AppComponent implements OnInit {
...
+          this.appVersionService.checkAppSupportedVersion(deviceInfo);
+        });
 
       // Global cache config
       GlobalCacheConfig.maxAge = 10 * 60 * 1000;
       GlobalCacheConfig.maxCacheCount = 100;
     });
   }
 
-  checkAppSupportedVersion() {
```
Comment (yksolanki9) [inline src/app/app.component.ts:147]: Moved this method inside `app-version` service and added it inside switch org as well.

File: src/app/core/services/offline.service.ts [hunk 2]
```diff
@@ -511,7 +509,10 @@ export class OfflineService {
     const delegatedAccounts$ = this.getDelegatedAccounts();
     const taxGroups$ = this.getEnabledTaxGroups();
 
-    this.loadAppVersion();
+    this.deviceService.getDeviceInfo().subscribe((deviceInfo) => {
+      this.loadAppVersion(deviceInfo);
+      this.appVersionService.checkAppSupportedVersion(deviceInfo);
```
Comment (yksolanki9) [inline src/app/core/services/offline.service.ts:514]: Added check for supported version in switch org page as well. This fixes a corner case where user is able to login to an deprecated app and create expense

File: src/app/app.component.ts [hunk 3]
```diff
@@ -189,7 +169,6 @@ export class AppComponent implements OnInit {
         .forEach((key) => lstorage.removeItem(key));
     }
 
-    this.checkAppSupportedVersion();
```
Comment (yksolanki9) [inline src/app/app.component.ts:192]: Moved this method inside `platform.ready()` as capacitor plugins may not be always initialized inside `ngOnInit`

File: src/app/app.component.ts [hunk 4]
```diff
@@ -136,43 +137,22 @@ export class AppComponent implements OnInit {
       });
       setTimeout(async () => await SplashScreen.hide(), 1000);
 
-      this.offlineService.loadAppVersion();
+      from(this.routerAuthService.isLoggedIn())
+        .pipe(
+          filter((isLoggedIn) => !!isLoggedIn),
```
Comment (yksolanki9) [inline src/app/app.component.ts:142]: `/mobileapp/check` and `version/app/<os>` calls should only be made if the user is logged in. Else these checks will be done  after login in switch org page

Comment (rvab) [reply src/app/app.component.ts:142]: Why these calls should be made after the user is logged in? What happens if done before and Why?

**Review by rvab — CHANGES_REQUESTED** [review]

File: src/app/app.component.ts (hunk 4 above)
Comment (yksolanki9) [reply src/app/app.component.ts:142]: I have already added the sentry issue link in the description. We'd get `403: Forbidden` as the user does not have a valid access token

File: src/app/core/services/app-version.service.ts [hunk 5]
```diff
@@ -43,35 +50,52 @@ export class AppVersionService {
...
+          lastLoggedInVersion,
+          user_email: eou?.us?.email,
+          appVersion: deviceInfo.appVersion,
+        });
+        this.router.navigate(['/', 'auth', 'app_version', { message: appSupportDetails.message }]);
+      }
+    });
+  }
+
   isSupported(data) {
```
Comment (Julias0) [inline src/app/core/services/app-version.service.ts:99]: isSupported(deviceInfo) {

File: src/app/core/services/app-version.service.ts [hunk 6]
```diff
@@ -43,35 +50,52 @@ export class AppVersionService {
...
+      lastLoggedInVersion: this.loginInfoService.getLastLoggedInVersion(),
+      eou: from(this.authService.getEou()),
+    }).subscribe(({ appSupportDetails, lastLoggedInVersion, eou }) => {
+      if (!appSupportDetails.supported && environment.production) {
+        this.trackingService.eventTrack('Auto Logged out', {
+          lastLoggedInVersion,
+          user_email: eou?.us?.email,
+          appVersion: deviceInfo.appVersion,
+        });
+        this.router.navigate(['/', 'auth', 'app_version', { message: appSupportDetails.message }]);
```
Comment (Julias0) [inline src/app/core/services/app-version.service.ts:94]: we shouldnt inject the router

Comment (Julias0) [reply src/app/core/services/app-version.service.ts:94]: Move subscribe block to page level component

File: src/app/core/services/offline.service.ts [hunk 7]
```diff
@@ -511,7 +509,10 @@ export class OfflineService {
     const delegatedAccounts$ = this.getDelegatedAccounts();
     const taxGroups$ = this.getEnabledTaxGroups();
 
-    this.loadAppVersion();
+    this.deviceService.getDeviceInfo().subscribe((deviceInfo) => {
```
Comment (Julias0) [inline src/app/core/services/offline.service.ts:512]: move the behaviour change to inside switch org page

**Review by Julias0 — APPROVED** [review]

File: src/app/app.component.ts (hunk 4 above)
Comment (rvab) [reply src/app/app.component.ts:142]: Why does the user need to be authorized to make this API? This check should be independent no? If the app version is deprecated, it's deprecated. Why depend on login status?

**Review by rvab — APPROVED** [review]
One clarification

File: src/app/app.component.ts (hunk 4 above)
Comment (yksolanki9) [reply src/app/app.component.ts:142]: I dont think we can allow any of our internal apis to be accessible without a token. Anyone can use a DDoS attack against us?

File: src/app/core/services/app-version.service.ts (hunk 5 above)
Comment (yksolanki9) [reply src/app/core/services/app-version.service.ts]: Fixed

File: src/app/core/services/app-version.service.ts (hunk 6 above)
Comment (yksolanki9) [reply src/app/core/services/app-version.service.ts]: Done

File: src/app/core/services/offline.service.ts (hunk 7 above)
Comment (yksolanki9) [reply src/app/core/services/offline.service.ts]: Done

**Pull request merged**